### PR TITLE
feat: Multiple Passkeys Support (#125)

### PR DESCRIPTION
**Pull request by hadv**
## Overview

Implements #125 - Adds support for multiple passkeys per P256Account, allowing users to register passkeys on multiple devices for enhanced security and device redundancy.

### 1. Multiple Passkey Storage
- Each account can have **unlimited active passkeys** (no MAX_PASSKEYS limit)
- Passkeys stored with unique IDs: `passkeyId = keccak256(abi.encodePacked(qx, qy))`
- Metadata tracking: coordinates (qx, qy), deviceId, timestamp added, active status

### 2. Security Model
- ✅ **OR Logic**: Any active passkey can authorize transactions
- ✅ **Immediate operations**: No timelock for passkey add/remove (timelocks removed for better UX)
- ✅ **Cannot remove last passkey when 2FA enabled** (prevents account lockout)
- ✅ **EntryPoint-only access** (requires passkey signature)
- ✅ **Guardian recovery with 24-hour timelock** (the real security mechanism)

#### Adding Passkeys
```solidity
function addPasskey(bytes32 _qx, bytes32 _qy, bytes32 _deviceId) external
```
- Only callable via EntryPoint (requires existing passkey signature)
- Validates coordinates are non-zero
- No MAX_PASSKEYS limit (unlimited passkeys)
- Emits `PasskeyAdded` event

#### Removing Passkeys
```solidity
function removePasskey(bytes32 _qx, bytes32 _qy) external
```
- Only callable via EntryPoint (requires passkey signature)
- Immediate removal (no timelock)
- Prevents removing last passkey when 2FA is enabled
- Marks passkey as inactive and removes from array
- Emits `PasskeyRemoved` event

#### validateUserOp
- Checks WebAuthn signature against **ALL active passkeys** (OR logic)
- During counterfactual deployment: uses initial passkey from initCode
- After deployment: iterates through all active passkeys
- First valid signature passes verification

#### isValidSignature (EIP-1271)
- Checks P256 signature against **ALL active passkeys** (OR logic)
- Fallback to deprecated qx/qy for backward compatibility

### 5. View Functions
- `getActivePasskeyCount()`: Number of active passkeys
- `getPasskeyByIndex(uint256)`: Get passkey by array index
- `getPasskeyById(bytes32)`: Get passkey by ID
- `getAllPasskeys()`: Get all passkeys with metadata

### 6. Recovery Integration
- Guardian recovery **deactivates ALL existing passkeys**
- Adds new recovery passkey as the only active passkey
- Prevents compromised passkeys from being used after recovery

### ✅ Implemented Security Features

1. **No Timelock for Passkeys**
   - Passkey timelocks removed because they create false security
   - Attacker with compromised credentials can cancel user's defensive actions
   - Real defense: Guardian recovery with 24-hour timelock

2. **Last Passkey Protection**
   - Cannot remove the last active passkey when 2FA is enabled
   - Prevents account lockout

3. **Access Control**
   - All passkey operations require EntryPoint authentication
   - Only passkey signatures can add/remove passkeys (owner cannot bypass)

4. **Gas Optimization**
   - No MAX_PASSKEYS limit (users can add as many as needed)
   - Early exit in loops when signature is found
   - Swap-and-pop pattern for gas-efficient array removal

5. **Backward Compatibility**
   - Deprecated qx/qy storage maintained
   - Existing accounts work without changes
   - isValidSignature has fallback to old storage

### 🔒 Security Best Practices

1. **Passkey ID Generation**: Uses `keccak256(qx, qy)` for deterministic, collision-resistant IDs
2. **Active Flag**: Allows deactivation without losing history
3. **Metadata Tracking**: Stores timestamp and deviceId for audit trail
4. **Event Emission**: All operations emit events for transparency

### Smart Contract (`src/P256Account.sol`)
- ✅ Added multiple passkey storage structure
- ✅ Implemented immediate passkey add/remove (no timelock)
- ✅ Updated signature verification to check all active passkeys
- ✅ Integrated with guardian recovery
- ✅ Added comprehensive view functions
- ✅ Removed ~300 lines of timelock code
- ✅ All 50 tests passing (37 existing + 13 new)

### SDK (`frontend/src/lib/P256AccountSDK.js`)
- ✅ Added `addPasskey()` function
- ✅ Added `removePasskey()` function
- ✅ Added `enableTwoFactor()` function
- ✅ Added `disableTwoFactor()` function
- ✅ Refactored `executeCall()` with callback pattern for 2FA
- ✅ Full support for owner-first signature flow

### UI Components (Clean V2 Implementations)
- ✅ **DeviceManagement.jsx**: Merge local and on-chain passkeys, immediate removal
- ✅ **PasskeySettings.jsx**: 2FA toggle with owner-first signature flow
- ✅ **AddCurrentDevice.jsx**: Add passkey on current device with immediate on-chain addition
- ✅ **AddMobileDevice.jsx**: Add passkey from mobile via QR code with immediate on-chain addition
- ✅ All components support 2FA mode (owner signature first, then passkey)

### CSS Cleanup
- ✅ Removed 54 lines of unused timelock-related CSS
- ✅ Made AddCurrentDevice.css and AddMobileDevice.css self-contained
- ✅ All CSS files are 100% efficient (no unused classes)

## Code Reduction Summary 📊

| Category | Before | After | Reduction |
|----------|--------|-------|----------|
| **Smart Contract** | ~800 lines | ~500 lines | **-300 lines (37%)** |
| **UI Components** | ~2,500 lines | ~1,200 lines | **-1,300 lines (52%)** |
| **CSS** | ~865 lines | ~811 lines | **-54 lines (6%)** |
| **Total** | ~4,165 lines | ~2,511 lines | **-1,654 lines (40%)** |

## Testing Status

✅ **All tests passing** (50 tests total)
- 37 existing tests (all passing)
- 13 new tests for multiple passkeys:
  - `test_AddPasskey()` - Verify adding a new passkey
  - `test_AddMultiplePasskeys()` - Add multiple passkeys
  - `test_RemovePasskey()` - Verify removing a passkey
  - `test_CannotRemoveLastPasskeyWhen2FAEnabled()` - 2FA protection
  - `test_CanRemoveLastPasskeyWhen2FADisabled()` - Allow when 2FA off
  - `test_ValidateUserOpWithMultiplePasskeys()` - OR logic verification
  - `test_ValidateUserOpWithSecondPasskey()` - Second passkey works
  - `test_ValidateUserOpWithInactivePasskey()` - Inactive passkey fails
  - `test_GetPasskeyFunctions()` - View functions work correctly
  - `test_AddPasskeyDuplicateCoordinates()` - Prevent duplicates
  - `test_RemoveNonexistentPasskey()` - Error handling
  - `test_GuardianRecoveryDeactivatesAllPasskeys()` - Recovery integration
  - `test_AddPasskeyAfterRecovery()` - Can add passkeys after recovery

## Commits Summary

1. `a9fb109` - Remove timelock from passkey operations
2. `e9073e0` - Remove duplicate getPasskeyCount() function
3. `4efe953` - Add comprehensive tests for multiple passkeys
4. `87926c7` - Add SDK functions for multiple passkeys management
5. `4e0d850` - Update DeviceManagement to merge local and on-chain passkeys
6. `b39e9fd` - Replace PasskeySettings with clean V2 implementation
7. `5625037` - Replace AddCurrentDevice with clean V2 implementation
8. `b3972a2` - Implement UserOperation-based passkey management and 2FA toggle
9. `86bb61b` - Implement 2FA signature flow with owner-first pattern
10. `2c5afeb` - Remove old component files (PasskeySettings, AddCurrentDevice)
11. `c45c376` - Replace AddMobileDevice with clean V2 implementation
12. `9dba7e6` - Remove old AddMobileDevice component
13. `7a566b1` - Cleanup unused CSS classes from timelock pattern
14. `1cb5180` - Make AddCurrentDevice and AddMobileDevice CSS self-contained

**Total:** 14 commits, all tests passing ✅

## Checklist

- [x] Smart contract implementation complete
- [x] Compilation successful
- [x] Security best practices applied
- [x] Documentation added
- [x] Tests updated (50 tests passing)
- [x] UI integration complete
- [x] SDK implementation complete
- [x] 2FA support implemented
- [x] CSS cleanup complete

---

## Ready for Review! 🎉

**All features implemented and tested:**
- ✅ Multiple passkeys support (unlimited)
- ✅ Immediate add/remove operations (no timelock)
- ✅ Full 2FA support with owner-first signature flow
- ✅ Clean V2 UI components
- ✅ Comprehensive test coverage (50 tests)
- ✅ 40% code reduction

**Please review the implementation with focus on:**
- Security model (no passkey timelock, guardian recovery is the defense)
- 2FA signature flow (owner first, then passkey)
- UI/UX improvements (immediate operations, clean components)
- Code quality and maintainability

---
Pull Request opened by [Augment Code](https://www.augmentcode.com/) with guidance from the PR author

### CODE REVIEW

**Comment by hadv** [comment]
## ⚡ Gas Optimization Update

I've added a significant gas optimization based on feedback:

### **Client-Specified PasskeyId**

Instead of looping through all passkeys during signature verification, the client now specifies which passkey they're using via `passkeyId` in the signature.

### **Updated Signature Format**

**Before:**
```
webAuthnSig || ownerSig(65)
```
- Contract loops through all passkeys (O(n))
- Worst case: 10 passkeys × ~100k gas = ~1M gas

**After:**
```
webAuthnSig || passkeyId(32) || ownerSig(65)
```
- Contract does direct lookup (O(1))
- Constant: ~100k gas regardless of passkey count
- **Gas savings: 0-900k gas** depending on number of passkeys

### **Backward Compatibility**

`isValidSignature` (EIP-1271) supports both formats:
- **Optimized**: `r(32) || s(32) || passkeyId(32)` - 96 bytes (O(1) lookup)
- **Legacy**: `r(32) || s(32)` - 64 bytes (loops for backward compatibility)

### **Security**

- Client cannot fake passkeyId - signature must still be valid for that specific passkey
- If passkeyId doesn't exist or is inactive, verification fails
- Same security guarantees, just more gas efficient

### **Implementation Details**

See commit: 511dfca

**Changes:**
- Updated `_validateSignature()` to extract and use passkeyId
- Updated `isValidSignature()` to support both 64-byte and 96-byte formats
- Updated documentation to reflect gas optimization
- Minimum signature length increased from 192 to 224 bytes for 2FA mode

File: src/P256Account.sol [hunk 1]
```diff
@@ -188,8 +242,15 @@ contract P256Account is IAccount, IERC1271, Ownable, Initializable {
             require(_qx != bytes32(0) && _qy != bytes32(0), "2FA requires passkey");
         }
 
+        // Set deprecated storage for backward compatibility
```
Comment (hadv) [inline src/P256Account.sol]: hmm, we're still in dev mode then don't need migration or backward compatibility

File: src/P256Account.sol [hunk 2]
```diff
@@ -32,10 +32,12 @@ contract P256Account is IAccount, IERC1271, Ownable, Initializable {
                                 STORAGE
     //////////////////////////////////////////////////////////////*/
 
-    /// @notice The P-256 public key x-coordinate
+    /// @notice DEPRECATED: The P-256 public key x-coordinate (kept for backward compatibility)
```
Comment (hadv) [inline src/P256Account.sol]: hmm, we're still in dev mode then don't need migration or backward compatibility

File: src/P256Account.sol [hunk 3]
```diff
@@ -32,10 +32,12 @@ contract P256Account is IAccount, IERC1271, Ownable, Initializable {
...
+    /// @dev Use passkeys mapping instead
     bytes32 public qx;
 
-    /// @notice The P-256 public key y-coordinate
+    /// @notice DEPRECATED: The P-256 public key y-coordinate (kept for backward compatibility)
```
Comment (hadv) [inline src/P256Account.sol]: hmm, we're still in dev mode then don't need migration or backward compatibility

**Comment by hadv** [comment]
## 🔒 Security Improvement: Pagination Added

Added pagination to prevent DoS/gas exhaustion attacks:

### **Problem**

The `getAllPasskeys()` function could be vulnerable to:
- **DoS attacks**: Malicious users could add hundreds/thousands of passkeys
- **Gas exhaustion**: Looping through all passkeys could exceed block gas limit
- **Failed transactions**: Users unable to query their passkeys if list is too large

### **Solution**

Added `getPasskeys(offset, limit)` with built-in protections:

```solidity
function getPasskeys(uint256 offset, uint256 limit)
    external
    view
    returns (
        bytes32[] memory passkeyIdList,
        bytes32[] memory qxList,
        bytes32[] memory qyList,
        uint256[] memory addedAtList,
        bool[] memory activeList,
        uint256 total  // Total count for pagination UI
    )
```

**Features:**
- ✅ **50-item cap**: Limit automatically capped at 50 to prevent gas issues
- ✅ **Offset/limit pagination**: Standard pagination pattern
- ✅ **Total count**: Returns total for UI pagination
- ✅ **Empty array handling**: Returns empty arrays if offset beyond total
- ✅ **Backward compatible**: Kept `getAllPasskeys()` with deprecation warning

### **Usage Example**

```javascript
// Get first page (items 0-49)
const page1 = await account.getPasskeys(0, 50);

// Get second page (items 50-99)
const page2 = await account.getPasskeys(50, 50);

// Total passkeys
console.log(`Total: ${page1.total}`);
```

### **Tests**

Added comprehensive pagination tests:
- ✅ `test_GetPasskeysPagination()` - Tests offset/limit, edge cases
- ✅ `test_GetPasskeysLimitCap()` - Verifies 50-item cap enforcement

---

**Commit:** 81d2bfc

This ensures the contract remains secure even with unlimited passkeys! 🔒

File: src/P256Account.sol [hunk 4]
```diff
@@ -32,19 +32,47 @@ contract P256Account is IAccount, IERC1271, Ownable, Initializable {
...
     /// @notice Two-factor authentication enabled flag
     /// @dev When enabled, transactions require both P-256 passkey signature and owner ECDSA signature
     bool public twoFactorEnabled;
 
+    /*//////////////////////////////////////////////////////////////
+                          MULTIPLE PASSKEYS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Passkey information
+    struct PasskeyInfo {
```
Comment (hadv) [inline src/P256Account.sol]: can we add device name or id. but keep it in short

File: src/P256Account.sol [hunk 5]
```diff
@@ -458,6 +545,8 @@ contract P256Account is IAccount, IERC1271, Ownable, Initializable {
      * @notice Execute a pending public key update
      * @param actionHash The hash of the action to execute
      * @dev Can be executed by anyone after timelock expires
+     * @dev NOTE: This is deprecated in favor of addPasskey/removePasskey
```
Comment (hadv) [inline src/P256Account.sol]: we're in dev mode then I think we won't keep this

File: src/P256Account.sol
Comment (hadv) [inline src/P256Account.sol]: we're in dev mode then I think we won't keep this
